rfcs/text/0000-metadata-integration-test.md

```diff
@@ -0,0 +1,80 @@
+### DCP PR:
+
+# Metadata integration testing
+
+## Summary
+
+This document details an integration test design for metadata schema changes in the Human Cell Atlas (HCA) Data Coordination Platform (DCP).
+
+## Authors
+
+* [Andrey Kislyuk](mailto:akislyuk@chanzuckerberg.com)
+* [Sam Pierson](mailto:spierson@chanzuckerberg.com)
+* [Matt Weiden](mailto:mweiden@chanzuckerberg.com)
+* [Mark Diekhans](mailto:markd@ucsc.edu)
+* [Hannes Schmidt](mailto:hannes@ucsc.edu)
+
+## Shepherd
+
+[Norman Morrison](mailto:norman@ebi.ac.uk)
+
+## Motivation
+
+Currently breakages to DCP systems caused by metadata schema changes are detected at runtime upon data (bundle) upload. This forces data wranglers, DCP operators, and data consumers to react to unpredictable runtime failures rather than detecting problems before the code is merged and deployed.
```

Comment:
I appreciate the suggestions. Let me try to incorporate them point by point.

> bundles are uploaded to the DSS but not the DCP (they are generated by ingest from _submissions_)

I've removed the word `(bundles)`

> runtime failures aren't exactly unpredictable

My intent here was the was to describe the system as it is now, rather than describing what it could be. Further, I did not take the rewrite wholesale as I don't believe that it addresses the core concern of the RFC: under any of the proposed solutions to metadata schema evolution (schema contract assertions, metadata migrations, metadata "APIs"), we'll need an integration test its assumptions and to gauge impact on critical data consumers.

> what does "the code" refer to

Good catch, replacing this with `metadata schema changes`.

> why do data wranglers and data consumers need to react (wranglers have successfully handed off new data, consumers will get an error and therefore just fail to incorporate the newest data)?

Under the current system:

1. Data wranglers will have to check the progress of an (frequently unreliable) upload, re-upload data, or work with the dev team to get the upload through.
1. I'm not sure that I understand your point about data consumers. If we know that a metadata schema change might break an important data consumer, say bioconductor, we may choose to make the change differently or contact them to proactively adapt their code before we push a great deal of data to them. This seems to be consistent with the original text.

Reply:
> Currently, breakages to DCP systems caused by metadata schema changes are detected at runtime upon data upload

My previous comment was based on a broader interpretation (see "submissions making use of new experimental graph shapes may also trigger a breakage without any schema changes") of this RFC than I think perhaps you intend. If this is out of scope, this text is fine. Maybe update the title and the summary to reflect the narrower scope though?

> react to unpredictable runtime failures

I still don't like use of "unpredictable" here. Failures are very predictable - wranglers are investing a large chunk of time in researching which fields components are using to warn about upcoming changes, and the metadata change SOP builds time in for the various changes to allow for developer review and comment. Clearly this process isn't working, so I support better testing, but describing failures as unpredictable is factually incorrect IMO

> Data wranglers will have to check the progress of an (frequently unreliable) upload, re-upload data, or work with the dev team to get the upload through.

I'm not seeing that uploads are frequently unreliable due to metadata schema changes - aren't the vast majority of new schemas successfully integrated into ingest and the DSS? I don't have precise stats on where failures occur but I'm sure they're easy to obtain... there's a large amount of reingestion happening currently but my understanding is that this is due to scaling and representational problems, not unsuccessful upload. Is that not correct?

> data consumers. Failing to incorporate the newest data and then having to scramble to adapt before the DSS retry window elapses in order to get the new data

OK, this makes sense - we're using a different definition of data consumer. I'd consider a data consumer as an end user (possibly an application) but you're thinking specifically of developers of data consuming applications. Maybe state this explicitly?

Reply:
OK, I see. I've made the following changes, which, I hope, address your concerns.

1. Retitled the RFC "Integration testing for metadata schema changes" - while some systems, like Azul, do rely on the graph structure, I don't think we can reasonably require testing for every new structure, given that almost every experimental structure is unique.
1. Removed the word `unpredictable`
1. Removed the words `(frequently unreliable)`
1. Changed `data consumers` to `third-party developers`